### PR TITLE
Learning Rate Sweep: Test lr=3e-4 and lr=1e-4 with Current Architecture

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current learning rate (lr=2e-4 with Lion optimizer) was set early in the project when the architecture was simpler. Since then, we've added:
- Gap/Stagger Spatial Bias (GSB, +2 input dims to spatial_bias MLP)
- PCGrad gradient surgery (modifies effective gradient at every step)
- Aft-foil SRF head (additional parameters)
- Surface refinement head (additional parameters)
- Multiple augmentation pipelines

The optimal learning rate may have shifted with these additions. Lion optimizer papers suggest LR should be 3-10x higher than AdamW — our 2e-4 may be conservative for the current architecture complexity.

**Expected impact:** If lr=3e-4 works, faster convergence → more effective epochs in the 180-min window → better final metrics. If lr=1e-4 works better, the model may be in a sensitive region where slower convergence finds a flatter basin.

### NO CODE CHANGES NEEDED — just change the --lr flag.

**Config A: lr=3e-4 (50% increase), 2 seeds:**
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/lr-3e4-s42" --wandb_group phase6/lr-sweep \
  --lr 3e-4 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "edward/lr-3e4-s73")
```

**Config B: lr=1e-4 (50% decrease), 2 seeds:**
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/lr-1e4-s42" --wandb_group phase6/lr-sweep \
  --lr 1e-4 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "edward/lr-1e4-s73")
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| lr=3e-4 | 42 | | | | | |
| lr=3e-4 | 73 | | | | | |
| **lr=3e-4 avg** | — | | | | | |
| lr=1e-4 | 42 | | | | | |
| lr=1e-4 | 73 | | | | | |
| **lr=1e-4 avg** | — | | | | | |
| **Baseline (lr=2e-4)** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

Also report: training loss curves — does lr=3e-4 converge faster? Does lr=1e-4 converge more smoothly? Any signs of instability (loss spikes)?

## Baseline

Current baseline (PR #2130, GSB + PCGrad, lr=2e-4):

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.05      |
| p_oodc | < 7.70       |
| p_tan  | **< 28.60**  |
| p_re   | < 6.55       |

W&B: d7l91p0x (s42), j9btfx09 (s73)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Learning Rate Sweep

### Setup
- 4 runs: lr=3e-4 (×2 seeds) and lr=1e-4 (×2 seeds)
- All runs: `--ema_start_epoch 100 --ema_decay 0.999 --cosine_T_max 160 --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias` + standard baseline flags
- All runs reached epoch 146-147 before 180-min timeout

### Results — Surface MAE (Pressure)

| Run | p_in | p_oodc | p_tan | p_re | val/loss | W&B Run ID |
|-----|------|--------|-------|------|----------|------------|
| lr=3e-4 s42 | 13.2 | 8.1 | 29.8 | 6.7 | 0.392 | `3at64dy4` |
| lr=3e-4 s73 | 13.9 | 8.0 | 29.9 | 6.4 | 0.394 | `vohbzzhi` |
| **lr=3e-4 avg** | **13.55** | **8.05** | **29.85** | **6.55** | — | — |
| lr=1e-4 s42 | 13.7 | 8.2 | 29.6 | 6.6 | 0.402 | `8s56fget` |
| lr=1e-4 s73 | 14.0 | 8.2 | **29.0** | 6.8 | 0.401 | `pyhi4uri` |
| **lr=1e-4 avg** | **13.85** | **8.20** | **29.30** | **6.70** | — | — |
| **Baseline (lr=2e-4)** | **13.05** | **7.70** | **28.60** | **6.55** | — | — |

### Comparison vs Baseline (2-seed averages)

| Metric | lr=3e-4 | lr=1e-4 | Baseline (lr=2e-4) |
|--------|---------|---------|-------------------|
| p_in | 13.55 (+3.8%) | 13.85 (+6.1%) | 13.05 |
| p_oodc | 8.05 (+4.5%) | 8.20 (+6.5%) | 7.70 |
| p_tan | 29.85 (+4.4%) | 29.30 (+2.4%) | 28.60 |
| p_re | 6.55 (0%) | 6.70 (+2.3%) | 6.55 |

### Analysis

**Neither alternative LR beats the baseline lr=2e-4.** Both directions are worse on most metrics.

**lr=3e-4 (higher):**
- Slightly worse across the board. p_in and p_oodc regress ~4-5%.
- p_tan at 29.85 is 4.4% worse than baseline's 28.60.
- p_re matches baseline (6.55).
- The 50% LR increase doesn't help convergence — the model reaches about the same epoch count, but the larger step size appears to overshoot some basins.

**lr=1e-4 (lower):**
- Worst on p_in (13.85, +6.1%) and p_oodc (8.20, +6.5%).
- Best individual p_tan (29.0 for s73), but the 2-seed average (29.30) still doesn't beat baseline (28.60).
- Higher val/loss (0.401-0.402 vs baseline ~0.385), suggesting underfitting within the 180-min budget.
- Slower convergence means the cosine schedule doesn't match well — the model hasn't fully converged when the schedule ends.

**Training dynamics:**
- lr=3e-4: Similar convergence speed as baseline, slightly higher training loss variance.
- lr=1e-4: Visibly slower convergence; hasn't reached the loss plateau that lr=2e-4 achieves.

### Conclusion

**lr=2e-4 remains optimal for this architecture.** The current LR is well-tuned — neither a 50% increase nor decrease improves any metric on average. The cosine schedule (T_max=160) is calibrated for lr=2e-4's convergence speed; changing LR without adjusting T_max creates a mismatch.

### Suggested Follow-ups
1. **Close this direction** — lr=2e-4 is well-calibrated.
2. If revisiting, try lr=2.5e-4 or lr=1.5e-4 for finer resolution, but the gap between 1e-4 and 3e-4 suggests the optimum is at or very near 2e-4.
3. Consider **warmup** or **cyclical LR** rather than changing the base rate.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2149 because ADVISOR: **Closing — LR sweep confirms lr=2e-4 is optimal.**

Neither alternative beat baseline:
- lr=3e-4 avg: p_in=13.55 (+3.8%), p_oodc=8.05 (+4.5%), p_tan=29.85 (+4.4%), p_re=6.55 (0%)
- lr=1e-4 avg: p_in=13.85 (+6.1%), p_oodc=8.20 (+6.5%), p_tan=29.30 (+2.4%), p_re=6.70 (+2.3%)

The cosine schedule (T_max=160) is calibrated for lr=2e-4's convergence speed — changing LR without adjusting T_max creates a mismatch. lr=1e-4 underfits within the 180-min budget, lr=3e-4 overshoots.

Good suggestion on cyclical LR / warmup as follow-ups — those change the schedule shape without shifting the base rate.

lr=2e-4 now confirmed optimal. Adding to dead ends. Edward will be reassigned to a new experiment..
